PairZipped classes to combine two Iterables/Iterators into Pairs. #36

lemonboston opened the pull request:
@dmfs Question for when you have looked at the code:
Should `PairingFunction` be a top level class?
`PairMatcher` - create essentials-testing module and implement it there properly? (in separate ticket)

dmfs commented:
I don't see any reasons to keep `PairingFunction` as an inner class.

I'm not quite happy with the name `ZippedPairs`. It doesn't quite sound right. `PairZipped` seems more appropriate.

`PairMatcher` should indeed be in an essentials-testing module. We certainly could also use a `SingleMatcher`.

dmfs commented:
It's ok to keep `PairMatcher` as an inner class until we have the testing module. When you create the ticket, make sure to add that `PairMatcher` should delegate to other `Matchers` to test the values.

lemonboston commented:
I've made `PairingFunction` top level, renamed class to `PairZipped` and also added `Iterator` version.
Created test for that, too, but there are multiple matchers now there as inner classes:
PairMatcher, IteratorMatcher, IsEmptyIteratorMatcher. I'll create tickets for those.

dmfs commented on essentials/src/test/java/org/dmfs/jems/function/elementary/PairingFunctionTest.java:
```diff
@@ -0,0 +1,60 @@
+    @Test
+    public void test()
+    {
+        LeftHand leftHand = new LeftHand();
```
This test could be shorter with the `PairMatcher`.

dmfs replied:
Please create an issue to improve this test and `ValuePairTest` by using `PairMatcher`.

lemonboston replied:
Ok, created: https://github.com/dmfs/jems/issues/42

dmfs commented on essentials/src/test/java/org/dmfs/jems/function/elementary/PairingFunctionTest.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+
+    private static final class LeftHand
```
Why did you create these static classes? Wouldn't it be sufficient to create to `new Object()`? Does it add to the test if the values are of a specific type? Same goes for `ValuePairTest`.

lemonboston replied:
I think it can add to the correctness guarantees to the test if actual, different types are used at the place of generic types. To 'test' that part as well. Not that important when there is only one type parameter in the scope of the test, but even then it may add to the readability. (That's why I created that `Contract` class in contentpal testing).

So what I mean by correctness is that we could mess up type parameters as well in theory, but if everything is Object in the test, we wouldn't notice.
Maybe we could have GenericType, GenericType1, GenericType2 'dummy' types available for this purpose. Or something shorter.
